IHP/QueryBuilder.hs

```diff
@@ -826,6 +830,32 @@ filterWhereCaseInsensitive (name, value) queryBuilderProvider = injectQueryBuild
         queryBuilder = getQueryBuilder queryBuilderProvider
 {-# INLINE filterWhereCaseInsensitive #-}
 
+
+filterWhereIdIn :: forall table model queryBuilderProvider (joinRegister :: *). (KnownSymbol table, Table model, model ~ GetModelByTableName table, HasQueryBuilder queryBuilderProvider joinRegister) => [Id model] -> queryBuilderProvider table -> queryBuilderProvider table
+filterWhereIdIn values queryBuilderProvider =
+    -- TODO Null values are ignored here for now, because they need special treatment as in sql they must be compared using "IS NULL"...
+    --  We would a) need to know somehow which values are null (which is not possible with primaryKeyConditionForId returning opaque Actions)
+    --  and b) then decompose the values into something like: (col_a IS NULL AND (col_b, col_c) IN ?) OR (col_b IS NULL AND (col_a, col_c) IN ?)
```

Comment:
To resolve this I would change the `Table` typeclass from 
```haskell
primaryKeyConditionForId :: Id record -> [(Text, PG.Action)]
```
to
```haskell
primaryKeyConditionForId :: Id record -> [(Text, FilterOperator, PG.Action)]
```
and the implementation would get the `FilterOperator` from a call to `toEqOrIsOperator` on the value of the id.

However `Table` is defined in `ModelSupport`, whereas `FilterOperator` and `EqOrIsOperator` are defined in `QueryBuilder` which imports from `ModelSupport`, so in order to avoid cyclic dependencies I would either have to move them and everything they depend on to `ModelSupport` and then reexport from `QueryBuilder` to keep backwards compatibility, but it just seems to me that they don't really belong into `ModelSupport`. We could move them to a whole new `FilterSupport` or something like that...

Or of course I could just leave everything as is and leave this caveat in, as NULLs in primary columns are really a niche thing to do anyways

Whats the best way to progress here?

Reply:
> We could move them to a whole new FilterSupport or something like that...

Sounds good, let's move it to `IHP.Filter` 👍

Reply:
I just realized that NULLs in primary keys aren't even a "niche thing to do" but actually not possible, as "Primary Key" implies "Unique" as well as "Not Null"

https://www.postgresql.org/docs/current/ddl-constraints.html#DDL-CONSTRAINTS-PRIMARY-KEYS

So this is pretty much all solved already :D

Reply:
👍  TIL :D

Reply:
I thought I had seen nulls in PKs somewhere, but looking around it seems like its sql standard to not allow that... Must have imagined it :D